app/src/main/java/com/example/househomey/form/ItemFormFragment.java

```diff
@@ -179,4 +199,63 @@ public void afterTextChanged(Editable editable) {
             }
         });
     }
+
+    /**
+     * Initializes photo components for the fragment (photo adapter,
+     * and image picker dialog) as well as sets this fragment to the components' listeners.
+     *
+     * @param rootView The root view of the fragment.
+     */
+    private void initAddImageHandler(View rootView) {
+        // Set up adapter for gallery and listener for adding photos
+        photoAdapter = new PhotoAdapter(getContext(), photoUris);
+        ((RecyclerView) rootView.findViewById(R.id.add_photo_grid)).setAdapter(photoAdapter);
+        photoAdapter.setOnAddButtonClickListener(this);
+
+        // Add listener for when image is taken/selected
+        imagePickerDialog = new ImagePickerDialog();
+        imagePickerDialog.setOnImagePickedListener(this);
+    }
+
+    /**
+     * Callback method for when the Add button in the photo gallery is clicked.
+     */
+    @Override
+    public void onAddButtonClicked() {
+        imagePickerDialog.show(getParentFragmentManager(), imagePickerDialog.getTag());
+    }
+
+    /**
+     * Callback method for handling the image URI returned from image picker dialog.
+     *
+     * @param imageUri The URI of the selected/taken image.
+     */
+    @Override
+    public void onImagePicked(String imageUri) {
+        imagePickerDialog.dismiss();
+        photoUris.add(imageUri);
+        photoAdapter.notifyItemInserted(photoAdapter.getItemCount() - 1);
+    }
+
+    /**
+     * Uploads a local image to Firebase Cloud Storage and returns its UUID.
+     *
+     * @param imageUri The local URI of the image to be uploaded.
+     * @return The UUID of the uploaded image.
+     */
+    private String uploadImageToFirebase(String imageUri) {
+        if (imageUri.contains("content://") || imageUri.contains("file://")) {
+            // New image to upload, create a unique storage reference
+            String imageId = UUID.randomUUID().toString();
```

Comment:
It looks like Cloud Storage does have a way to automatically assign a UID when uploading a new file. 

BUT, the generated ID would be returned to us asynchronously. So we would have to add some asynch/await style code to wait for it to return the ID.

We need the ID to append it to the Item document's `photoIds` field, so we know which images the item references in Cloud Storage. So I think just generating it here is simpler than asynch/awaiting

Reply:
@ldbonkowski I made two review fixes for you:
1. found a way to remove both setListener methods
2. restructured Cloud Storage to have the dir structure `images/{username}/{image_UUID}`

Last point of debate on the Firestore vs. Cloud Storage topic. Currently, we are uploading the images before the document is made. So if the document fails, we will have images in the storage that can never be deleted by the user. I think this is less than ideal.

The best solution I can think of is moving the image upload method call within the Firestore document add/edit `onSuccessListener`. This will look a lot more ugly, but would only allow images to be created if doc succeeds. No need to consider if image uploads fails, because users will still be able to still delete that broken "image", even if its not loading.

Let me know if this is worth making the change. It will require an additional Firestore update to the document's photoIds field. So flow would look like `Add Item Document (w/o photoIds) -> Wait for Success -> Upload Images to Cloud -> Update Item Document with Resulting photoIds`. I believe this would allow us to let Firebase auto generate the photo ids too

Reply:
Awesome!

Only if you want to make this change. I am happy as is. 

I am pretty sure Firebase will only generate the photo id if we were to add it as a collection.